lib/events.js

```diff
@@ -75,15 +78,58 @@ EventEmitter.init = function() {
     }
   }
 
-  if (this._events === undefined ||
-      this._events === Object.getPrototypeOf(this)._events) {
-    this._events = Object.create(null);
-    this._eventsCount = 0;
-  }
+  this[kEventsProxy] = null;
+  if (this[kEvents] === Object.getPrototypeOf(this)[kEvents])
+    this[kEvents] = undefined;
+
+  Object.defineProperty(this, '_events', eventsProperty);
 
   this._maxListeners = this._maxListeners || undefined;
 };
 
+// Events proxy for compatibility with older user-land code
+const proxyEventsHandler = {
+  get({ emitter }, prop) {
+    let events = emitter[kEvents];
+    if (events === undefined)
+      events = emitter[kEvents] = new Map();
+    return events.get(prop);
+  },
+  getPrototypeOf() {
+    return null;
+  },
+  ownKeys({ emitter }) {
+    const events = emitter[kEvents];
+    if (events === undefined)
+      return [];
+    return [...events.keys()];
```

Comment:
Since `Object.keys()` uses `getOwnPropertyDescriptor` hook (or `Reflect.getOwnPropertyDescriptor` if there isn't one) to check if the returned keys are valid, and the hook isn't implemented, any attempt to call `Object.keys()` on `_events` would result in an empty array.

Reply:
Thanks. There's quite a few things (like returning true within `set`) I missed since I didn't write tests for the Proxy yet.